### PR TITLE
Add feature-variables API for mobile

### PR DESCRIPTION
**Pull request by jhugman**
[HTML Render][1]

[1]: https://github.com/mozilla/experimenter-docs/blob/f70c23a4339fde68628b2ec3af786dfa669ab566/docs/feature-variables-and-me.md

## Description

This PR adds a descriptive document for mobile product teams on using the Feature API.

## Issue that this pull request resolves (optional)

This will link to and close an issue in GitHub. Replace `experimenter` with the repository where the issue lives and replace `0000` with the issue number. Remove this section if not applicable.

Closes: mozilla/experimenter#0000

## Other information (optional)

Any other information or requests important to this pull request. Remove this section if not applicable.

If you're not certain that your Markdown will render correctly, you can include this line:
I have not ran the project locally with my changes and it would be be ideal for the reviewer to check into my branch and ensure images etc. are rendering as expected.

### CODE REVIEW

File: docs/feature-variables-and-me.md [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+The API used by software engineers is relatively small, so this document is not just for them. This document is also for designers, product managers and engineering managers who design, work on, and are responsible for features in the mobile apps we build. Successful experimentation requires multiple parts of the team to share an understanding of these concepts.
+
+> 👋 Information
+>
+> Neither Nimbus, nor all of the Feature Variables work are finished yet, but they are certainly still useful. This document will talk about features that aren't yet implemented, but will serve to illustrate the concepts. A sidebar like this will tell you when a feature isn't ready.
+
+> ⛅️🔬🔭 Nomenclature
+>
+> Much of the literature around the methodology of experiments Nimbus implements has its roots in medical testing. The feature variables API do not require understanding of double blind experiments or data-science, but this document will occasionally use words like "treatment" or "exposure".
```
Comment (eliserichards) [inline docs/feature-variables-and-me.md:15]:
```suggestion
> Much of the literature around the methodology of experiments Nimbus implements has its roots in medical testing. The feature variables API does not require understanding of double blind experiments or data-science, but this document will occasionally use words like "treatment" or "exposure".
```

File: docs/feature-variables-and-me.md [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+
+One easy way to start thinking about features, would be to identify user-visible surfaces of the app: the `new-tab` screen, the `app-menu`, the `context-menu`, the `onboarding`.
+
+However, there is one rule:
+
+> ⛅️🔬🔭 Concept
+>
+> *For a given user each feature can only be involved in one experiment at a time*.
+>
+> If we see user change behavior after being exposed to an experimental treatment we need to be able to attribute it to that treatment, not another from a different experiment.
```
Comment (eliserichards) [inline docs/feature-variables-and-me.md:29]: This is great context to have

File: docs/feature-variables-and-me.md [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+    "settings-menu-item-title": "Settings",
+    "settings-menu-item-icon": "ic_settings",
+    "settings-menu-item-enabled": true,
+    "settings-menu-item-action": "firefox://settings"
+}
+```
+
+This JSON object looks like what experimenters will be putting into branch configuration screens in Experimenter, under Feature Configuration.
+
+Where did these keys come from? This is not up to Nimbus, but up to the app, i.e. the app team. In this hypothetical case, you have some theories about the title and the icon, and now the app needs to get those values from nimbus.
```
Comment (eliserichards) [inline docs/feature-variables-and-me.md:52]: By "the app team" do you mean the engineers who will be ultimately implementing the design (in this example)? Or would "team" in this context still mean _anyone_ on the team?

Comment (eliserichards) [reply docs/feature-variables-and-me.md:52]: (I don't think this sentence necessarily has to change, I'm just curious)

File: docs/feature-variables-and-me.md [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+### Fundamental types: Strings, Int, Bool
+
+`getString()`, `getBool()` and `getInt()` all return exactly is in the JSON. If there is a disagreement about types, i.e. if the app is expecting a string, and the value in the JSON is an integer, the app gets `nil` or `null`.
+
+### Everything is optional
+
+If the app asks for a variable that is not specified in this particular experiment, then it gets back `nil` or `null`.
+
+It is thus imperative that the app has a reasonable default. On the other hand, this allows us to have experiments which configure only small parts of a feature.
```
Comment (eliserichards) [inline docs/feature-variables-and-me.md:84]: 👏

File: docs/feature-variables-and-me.md [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+
+For example, on Android: `getText("settings-menu-item-title")` may get a string from the JSON `"app_menu_settings_title"`, which is then resolves to `R.string.app_menu_settings_title` which is then used to look up the String in the `Resources`.
+
+On iOS, `getText` uses a similar process via `LocalizedString` to look up the translated strings. You can specify the `tableName` as well as the `key` in the single value by joining it with a slash.
+
+For example, `getText("settings-menu-item-title")` may get a string from the JSON `"AppMenu/SettingsTitle"` which uses `bundle.localizedString("SettingsTitle", tableName: "AppMenu")` to look up a localized string. If the app doesn't use `tableName`, then you can omit it: e.g. `AppMenu_SettingsTitle` would look for `NSLocalizedString("AppMenu_SettingsTitle")`.
+
+If `getString()` returned a string, and the resource lookup didn't succeed, `getText()` falls back to the string.
+
+This means that you can use either pre-translated strings to try out experiments across locales, or target your experiment on a single language.
```
Comment (eliserichards) [inline docs/feature-variables-and-me.md:98]: 👏

File: docs/feature-variables-and-me.md [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+
+> 🎛 Configuration
+>
+> Resource lookup via `Bundle` and `Context` uses the objects passed to nimbus at construction time at app-startup. In Firefox for iOS and Fenix this is `Bundle.main` and `context.applicationContext` respectively.
+
+### Image resources
```
Comment (eliserichards) [inline docs/feature-variables-and-me.md:104]:
```suggestion
### 📷 Image resources
```

File: docs/feature-variables-and-me.md [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+
+In the example above, the `icon` uses `getImage()` and its Android analog `getDrawable()`. This gets a string value from the JSON with `getString()` and then uses that value to look up the pre-bundled resource.
+
+For example on Android: `getDrawable("settings-menu-item-icon")` may get a string `"ic_settings"`, which is then resolved to `R.drawable.ic_settings`, which is then resolved to `context.resources.getDrawable(R.drawable.ic_settings)`.
```
Comment (eliserichards) [inline docs/feature-variables-and-me.md:108]: Should this be:
`getDrawable("@drawable/settings-menu-item-icon")`
?

File: docs/feature-variables-and-me.md [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+
+On iOS: `getImage("settings-menu-item-icon")` may get a string `"icon_photon_gear"`, which is then used to get the named `UIImage` with `UIImage(named:in:)`.
+
+## Making JSON more manageable
+
+We focused on the settings menu item in the above example, as a way of making a small enough example to reason about in this documentation, but it made for some very long variable names. The `Variables` object has itself a `getVariables(key: String)` method to make navigating the JSON more easily. This in turn allows the JSON to be organized in different ways.
```
Comment (eliserichards) [inline docs/feature-variables-and-me.md:114]:
```suggestion
We focused on the settings menu item in the above example, as a way of making a small enough example to reason about in this documentation, but it made for some very long variable names. The `Variables` object has itself a `getVariables(key: String)` method to make navigating the JSON easier. This in turn allows the JSON to be organized in different ways.
```

Comment (eliserichards) [reply docs/feature-variables-and-me.md:114]: Is it possible to get an example of "organized in different ways"?

File: docs/feature-variables-and-me.md [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+}
+```
+
+> 👋 Information
+>
+> `variables.getVariables()` can be arbitrarily deep. `variables.getVariables()` returns an optional `Variables` object.
+
+## Recording exposure events
+
+> ⛅️🔬🔭 Enrollment versus Exposure
```
Comment (eliserichards) [inline docs/feature-variables-and-me.md:172]: This is something I have personally been thinking about lately. Love this.

File: docs/feature-variables-and-me.md [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+>
+> When a client is selected to take part in an experiment, they are _assigned_ a branch. This is _enrollment_.
+> However, the user may not be _exposed_ to the branch until sometime later. The exposure is the earliest moment that the user could be affected by the experimental treatment.
+>
+> Nimbus tracks the enrollment and the exposure events using Glean.
+
+For experiments in Firefox for iOS and Android, enrollment happens shortly after app-startup.
```
Comment (eliserichards) [inline docs/feature-variables-and-me.md:179]: Do enrollment and exposure only happen once for each user? (per experiment)

File: docs/feature-variables-and-me.md [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+    nimbus.recordExposureEvent("app-menu")
+    viewController.present(menuSheet, animated: true, completion: nil)
+}
+```
+
+## Using configurable features to experiment with another
+
+The feature itself may be configurable, but we don't have to limit feature configuration to experiments _about that feature_.
+
+We can imagine world where we have multiple features, say: an `app-menu`, `onboarding` and `newtab`. On each of these features we have a messaging surface, and we want to run an experiment to find which is the best surface to show the message about a behavior we wish to maximise: setting the browser to be the device default.
```
Comment (eliserichards) [inline docs/feature-variables-and-me.md:222]:
```suggestion
We can imagine a world where we have multiple features, say: an `app-menu`, `onboarding` and `newtab`. On each of these features we have a messaging surface, and we want to run an experiment to find which is the best surface to show the message about a behavior we wish to maximise: setting the browser to be the device default.
```

File: docs/feature-variables-and-me.md [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+>
+> While for any given user a feature may be involved in only one experiment, one experiment should be able to configure multiple features.
+
+Because both product teams' experiments require the `onboarding` experiment, no user will be involved in _both_ experiments.
+
+> 👋 Unimplemented
+>
+> *Coming soon* Configuring multiple features per branch isn't yet in Experimenter.
+
+## Working with configurable features
```
Comment (eliserichards) [inline docs/feature-variables-and-me.md:246]:
```suggestion
## 🔧⚙️ Working with configurable features
```

File: docs/feature-variables-and-me.md [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+
+Throughout the process of designing and building these configurable features, the feature variables have needed to be documented. At first, when the feature is being envisioned and designed, the variables should travel with the designs themselves.
+
+When the feature is being implemented, these variables will begin to acquire concrete names, types and organization, which will be used extract JSON from the Nimbus SDK and configure the application features themselves. This documentation will begin to take shape and textual organization that travel in the app's repository.
+
+When the feature is being tested, QA testers are going to want to configure the features within bounds and tolerances set by the designs and the engineers.
+
+Finally, when the feature is part of experiments, then the experiment owner, setting the branches in Experimenter needs to be able configure the branches with variables with spellings and organization that match the app.
```
Comment (eliserichards) [inline docs/feature-variables-and-me.md:254]:
```suggestion
Finally, when the feature is part of experiments, then the experiment owner setting the branches in Experimenter needs to be able configure the branches with variables with spellings and organization that match the app implementation.
```

File: docs/feature-variables-and-me.md [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+
+> 👋 Unimplemented
+>
+> The Nimbus team has a few ideas about this, for documenting the features in an app in a machine readable format, tentantively called the Manifest. We're thinking:
+> 
+> 1. generating code for app developers to use in configuration.
+> 2. generating UI for experimenters to use in experiment design
+> 3. generating UI for QA to use in feature testing and local development.
+>
+> We're in the early stages of planning this, so would love to hear some feedback or ideas you may have.
```
Comment (eliserichards) [inline docs/feature-variables-and-me.md:266]: This would be extremely helpful. We are experimenting with this currently, so let's talk more.

**Review by eliserichards — COMMENTED** [review]
I added a few typos and a thoughts. Looks good overall 👍 I think this will be really helpful.

File: docs/feature-variables-and-me.md (hunk 3 above)
Comment (jhugman) [reply docs/feature-variables-and-me.md:52]: Where do the variables come from? Who names them? They're part of the design, and then become part of the implementation. 

My intention was to convey that Nimbus does not take a view on how you arrange the JSON or spell the keys.

File: docs/feature-variables-and-me.md [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+val menuButton = Button(
+    icon = R.drawable.ic_menu,
+    onButtonPressed = {
+        nimbus.recordExposureEvent("app-menu")
+        show(menu)
+    }
+)
+```
+
+This is a characature of the same code in Swift.
```
[the remark this answers is not in the document]
Comment (jhugman) [reply docs/feature-variables-and-me.md:202]: oh, TIL.

**Review by travis79 — APPROVED** [review]
Nicely done!

File: docs/feature-variables-and-me.md (hunk 7 above)
Comment (jhugman) [reply docs/feature-variables-and-me.md:112]: `getDrawable("settings-menu-item-icon")` uses `getString("settings-menu-item-icon")`. I changed the text here to make that clearer.

File: docs/feature-variables-and-me.md (hunk 8 above)
Comment (jhugman) [reply docs/feature-variables-and-me.md:118]: Added a linking sentence to the example below.

File: docs/feature-variables-and-me.md (hunk 10 above)
Comment (jhugman) [reply docs/feature-variables-and-me.md:188]: Added an additional sentence to the concept-box above.

**Review by eliserichards — APPROVED** [review]
Great job! 🙌 🚀

**Pull request merged**